services/QuillLMS/app/queries/teachers_data.rb

```diff
@@ -16,30 +16,82 @@ def self.run(teacher_ids)
-      INNER JOIN (SELECT users.id AS teacher_id, COUNT(DISTINCT concept_results.id) AS number_of_questions_completed, activity_sessions.id AS activity_session_id FROM users
+
+    number_of_students_query = User.find_by_sql(
+      "SELECT
+        users.id,
+        users.name,
+        users.email,
+        COUNT(DISTINCT students_classrooms.id) AS number_of_students
+      FROM users
+      LEFT OUTER JOIN classrooms_teachers ON users.id = classrooms_teachers.user_id
+      LEFT OUTER JOIN classrooms ON classrooms_teachers.classroom_id = classrooms.id
+      LEFT OUTER JOIN students_classrooms ON classrooms.id = students_classrooms.classroom_id
+      WHERE users.id IN (#{teacher_ids_str})
+      GROUP BY users.id"
+    )
+
+    question_count_query = User.find_by_sql(
+      "SELECT 
+        users.id,
+        COUNT(DISTINCT concept_results.id) AS number_of_questions_completed
+      FROM users 
       INNER JOIN units ON users.id = units.user_id
       INNER JOIN classroom_units ON units.id = classroom_units.unit_id
       INNER JOIN activity_sessions ON classroom_units.id = activity_sessions.classroom_unit_id
       INNER JOIN concept_results ON activity_sessions.id = concept_results.activity_session_id
       WHERE users.id IN (#{teacher_ids_str})
       AND activity_sessions.state = 'finished'
-      GROUP BY users.id, activity_sessions.id) AS acss_ids ON activity_sessions.id = acss_ids.activity_session_id
-      GROUP BY acss_ids.teacher_id
-    ) AS time_spent_query ON users.id = time_spent_query.teacher_id
-    WHERE users.id IN (#{teacher_ids_str})
-    GROUP BY users.id, number_of_questions_completed")
+      GROUP BY users.id"
+    )
+
+    time_spent_query = User.find_by_sql(
+      "SELECT 
+        users.id,
+        #{time_spent} AS time_spent
+      FROM users 
+      INNER JOIN units ON users.id = units.user_id
+      INNER JOIN classroom_units ON units.id = classroom_units.unit_id
+      INNER JOIN activity_sessions ON classroom_units.id = activity_sessions.classroom_unit_id
+      WHERE users.id IN (#{teacher_ids_str})
+      AND activity_sessions.state = 'finished'
+      GROUP BY users.id"
+    )
+
+    combiner = {}
+    number_of_students_query.each do |row|
+      combiner[row.id] = {
+        name: row.name,
+        email: row.email, 
+        number_of_students: row.number_of_students
+      }
+    end
+
+    question_count_query.each do |row|
+      combiner[row.id][:number_of_questions_completed] = row.number_of_questions_completed
+    end
+
+    time_spent_query.each do |row|
+      combiner[row.id][:time_spent] = row.time_spent
+    end
+
+    result = combiner.keys.map do |key|
+      hash_value = combiner[key]
+      user = User.new(
+        id: key,
+        name: hash_value[:name],
+        email: hash_value[:email]
+      )
+      user.define_singleton_method(:number_of_students) do 
```

Comment:
This isn't as weird as it seems - `find_by_sql`, which was used previously, does this under the hood.

Reply:
This is a very interesting approach.